src/timed_action.h

```diff
@@ -10,114 +10,67 @@
-        CallbackWrapperData *wrapperData = reinterpret_cast<CallbackWrapperData *>(userData);
-        if (wrapperData->finishTime < std::chrono::steady_clock::now()) return JVMTI_VISIT_ABORT;
-        return reinterpret_cast<jvmtiHeapReferenceCallback>(wrapperData->callback)(refKind, refInfo, classTag, referrerClassTag, size, tagPtr, referrerTagPtr, length, wrapperData->userData);
-    }
-
-    static jint JNICALL iterateThroughHeapCallbackWrapper(jlong classTag, jlong size, jlong *tagPtr, jint length, void *userData) {
-        CallbackWrapperData *wrapperData = reinterpret_cast<CallbackWrapperData *>(userData);
-        if (wrapperData->finishTime < std::chrono::steady_clock::now()) return JVMTI_VISIT_ABORT;
-        return reinterpret_cast<jvmtiHeapIterationCallback>(wrapperData->callback)(classTag, size, tagPtr, length, wrapperData->userData);
-    }
+                                                        jlong *referrerTagPtr, jint length, void *userData);
+
+    static jint JNICALL iterateThroughHeapCallbackWrapper(jlong classTag, jlong size, jlong *tagPtr, jint length, void *userData);
 
     jvmtiError FollowReferences(jint heapFilter, jclass klass, jobject initialObject,
                                 jvmtiHeapReferenceCallback callback, void *userData,
-                                const char *debugMessage=nullptr) {
-        if (shouldStopExecution()) return MEMORY_AGENT_TIMEOUT_ERROR;
-
-        if (debugMessage != nullptr) {
-            debug(debugMessage);
-        }
-
-        jvmtiHeapCallbacks cb;
-        std::memset(&cb, 0, sizeof(jvmtiHeapCallbacks));
-        cb.heap_reference_callback = followReferencesCallbackWrapper;
-
-        CallbackWrapperData wrapperData(finishTime, reinterpret_cast<void *>(callback), userData);
-        return jvmti->FollowReferences(heapFilter, klass, initialObject, &cb, &wrapperData);
-    }
+                                const char *debugMessage=nullptr);
 
     jvmtiError IterateThroughHeap(jint heapFilter, jclass klass, jvmtiHeapIterationCallback callback,
-                                  void *userData, const char *debugMessage=nullptr) {
-        if (shouldStopExecution()) return MEMORY_AGENT_TIMEOUT_ERROR;
-
-        if (debugMessage != nullptr) {
-            debug(debugMessage);
-        }
-
-        jvmtiHeapCallbacks cb;
-        std::memset(&cb, 0, sizeof(jvmtiHeapCallbacks));
-        cb.heap_iteration_callback = iterateThroughHeapCallbackWrapper;
-
-        CallbackWrapperData wrapperData(finishTime, reinterpret_cast<void *>(callback), userData);
-        return jvmti->IterateThroughHeap(heapFilter, klass, &cb, &wrapperData);
-    }
+                                  void *userData, const char *debugMessage=nullptr);
 
 protected:
     JNIEnv *env;
     jvmtiEnv *jvmti;
+    std::string cancellationFileName;
     std::chrono::steady_clock::time_point finishTime;
 };
 
+
+#include "timed_action.hpp"
```

Comment:
What's the purpose of this?

Reply:
This technique is used to separate the template class declaration from implementation. A header with implementations is included at the bottom of the file because member template functions have to be implemented in the same file, otherwise, there will be a link error.